homeassistant/components/light/__init__.py

```diff
@@ -336,17 +340,24 @@ async def async_handle_light_on_service(light, call):
         if supported_color_modes:
             params.pop(ATTR_WHITE_VALUE, None)
 
-        # Zero brightness: Light will be turned off
         if params.get(ATTR_BRIGHTNESS) == 0:
-            await light.async_turn_off(**filter_turn_off_params(params))
+            await async_handle_light_off_service(light, call)
         else:
             await light.async_turn_on(**params)
 
+    async def async_handle_light_off_service(light, call):
+        """Handle turning off a light."""
+        params = dict(call.data["params"])
+
+        if ATTR_TRANSITION not in params:
+            profiles.apply_default(light.entity_id, True, params)
```

Comment:
What happens with the transition if we use the toggle service?

Reply:
My understanding is that the toggle service already relies on the light.is_on state. It just process' the the service call to be a light.turn_on or light.turn_off appropriately.

Reply:
That is not what I meant with this question. My question was on the transition itself.

This will call the off/on directly after each other. Now profile (and thus transition) becomes part of the turn off, does that mean it waits until the transition has been completed for turning off? Or?

Reply:
If a light has a default transition of 5 seconds, it would be treated the same as if `light.toggle`, `light.turn_on` or `light.turn_off` was called with a transition of 5 seconds. In other words, for this same light, calling `light.turn_off` with a transition time of 5 seconds and `light.turn_off` wihtout a specified transition time should both have identical results and behaviour.

Reply:
> it would be treated the same as if light.toggle, light.turn_on or light.turn_off was called with a transition of 5 seconds.

Agreed, but not if called without a transition.

Assume this service, with a light that sets the transition to 5 seconds in its profile, the light if on before calling this service.

```yaml
- service: light.toggle
  target:
    entity_id: light.example
  data: {}
```

Before this change:
- Turn off immediately, no transition
- Turns on the light, using transition from profile

With this change:
- Turns off light, with a transition of 5 seconds...
- ?????
- Turns on the light, with a transition of 5 seconds

The ???? part is interesting and is definitely different now. What does happen there? Does this whole change take 10 seconds now? Or still 5?

Reply:
> The ???? part is interesting and is definitely different now. What does happen there? Does this whole change take 10 seconds now? Or still 5?

10 seconds. There should be no ????? step, unless I've done something terribly wrong in my code.

Before:
* Turn off immediately, no transition
* Turns on the light, using transition from profile (5 seconds)

After:
* Turns off light, using transition from profile (5 seconds)
* Turns on the light, using transition from profile (5 seconds)

This is the same as using this as a service call:
```yaml
- service: light.toggle
  target:
    entity_id: light.example
  data: {
    transition:5
  }
`